Feature/tsv110

qcjiang opened the pull request:
Add instructions for [tsv110](https://en.wikichip.org/wiki/hisilicon/microarchitectures/taishan_v110), measured with ibench

JanLJL commented:
Hi @qcjiang,

I looked over your PR, first of all thanks for participating!
There are still some open questions from my side regarding the machine file and things I need you to do before merging this into the main branch:
- "*Header*":
    - there is the attribute `scheduler_size: '*'` but the wildcard does not exist here. I changed it to `None`. Is that ok for you?
    - I could not find any information about the reorder buffer size on the internet (you added `128`). Can you validate this number?
    - You set the load throughput to `[1, '34'], [1, '012']` but in the instruction forms (as well as in the port diagram) you state a load only needs 1 cy on ports 6 or 7 (`[1, '67']`). I changed that accordingly and added the default load/store throughput; if this was not right in your eyes, please let me know.
- Load/Stores:
    - Since we don't model RAW dependencies so far, we assume a latency of `0 cy` for store instructions. I changed that to be consistent with our previous machine files
    - Many of your load and store instruction forms have `None` (`~`) as `offset` or `index`. For most Arm micro-architectures, we don't measure any difference in throughput and latency for basic or enhanced addressing mode; I therefore changed that to wildcard operands. Please let me know if you disagree with those changes.
    - There are some instruction forms including post-indexed memory addressing without having an `index` attribute. This doesn't make sense and I deleted those.
    - For instruction forms including pre-indexed or post-indexed memory addressing, you didn't take the additional arithmetic operation (increment) into account, was this intentional? I changed that accordingly and assumed the increment can happen on one of ports `0`, `1`, `2`. Can you validate this behavior with benchmarks?
    - I applied the same changes to `ldp` and `stp` instructions since they are normally implemented as two stores or two loads. If the TSV behaves differently (i.e., smarter) here, please revert those changes.
- We normally assume a "free" register renaming, i.e., register to register moves without conversion, for our model. I changed those instructions accordingly.
- For many instructions, you added a throughput number, but neither a port pressure nor a latency. This will not lead to sufficient result for an analysis. I tried to add some port pressure assignments based on your diagram and previous input (e.g., `add`), but obviously the latency (and honestly, also the port pressure) is something you need to validate via benchmarks...
- Some instruction forms (`fdiv`) don't have any information. Either add the missing data or remove them please.
- There should be a basic set of instructions, like branch instructions and binary operators (`and`, `or`, ...) in the machine file. I added branch instructions based on your diagram, please verify this is correct and also add some binary operands after benchmarking.

Let me know if you have any questions, happy to discuss those things with you!

qcjiang commented:
Hi Jan, thanks for your detailed explain, I apologize for the incomplete PR, it is generated automatically from ibench. Last week I have been working on repeating your work on A72, but for our architecture tsv110, there is no port mapping file. I have to learn to use PMEvo to generate port mapping file, I tried several times but the result is not so well, I have reached Fabian Ritter. We will be working on to improve the correctness, but consider the workload I'd better try another way maybe manually write the yml file. Again I apologize for the incomplete commit, I will make a correct PR as soon as possible.---- Replied Message ***@***.***>Date11/30/2021 01:50 ***@***.***> ***@***.******@***.***>SubjectRe: [RRZE-HPC/OSACA] Feature/tsv110 (PR #80)
Hi @qcjiang,
I looked over your PR, first of all thanks for participating!
There are still some open questions from my side regarding the machine file and things I need you to do before merging this into the main branch:

"Header":

there is the attribute scheduler_size: '*' but the wildcard does not exist here. I changed it to None. Is that ok for you?
I could not find any information about the reorder buffer size on the internet (you added 128). Can you validate this number?
You set the load throughput to [1, '34'], [1, '012'] but in the instruction forms (as well as in the port diagram) you state a load only needs 1 cy on ports 6 or 7 ([1, '67']). I changed that accordingly and added the default load/store throughput; if this was not right in your eyes, please let me know.


Load/Stores:

Since we don't model RAW dependencies so far, we assume a latency of 0 cy for store instructions. I changed that to be consistent with our previous machine files
Many of your load and store instruction forms have None (~) as offset or index. For most Arm micro-architectures, we don't measure any difference in throughput and latency for basic or enhanced addressing mode; I therefore changed that to wildcard operands. Please let me know if you disagree with those changes.
There are some instruction forms including post-indexed memory addressing without having an index attribute. This doesn't make sense and I deleted those.
For instruction forms including pre-indexed or post-indexed memory addressing, you didn't take the additional arithmetic operation (increment) into account, was this intentional? I changed that accordingly and assumed the increment can happen on one of ports 0, 1, 2. Can you validate this behavior with benchmarks?
I applied the same changes to ldp and stp instructions since they are normally implemented as two stores or two loads. If the TSV behaves differently (i.e., smarter) here, please revert those changes.


We normally assume a "free" register renaming, i.e., register to register moves without conversion, for our model. I changed those instructions accordingly.
For many instructions, you added a throughput number, but neither a port pressure nor a latency. This will not lead to sufficient result for an analysis. I tried to add some port pressure assignments based on your diagram and previous input (e.g., add), but obviously the latency (and honestly, also the port pressure) is something you need to validate via benchmarks...
Some instruction forms (fdiv) don't have any information. Either add the missing data or remove them please.
There should be a basic set of instructions, like branch instructions and binary operators (and, or, ...) in the machine file. I added branch instructions based on your diagram, please verify this is correct and also add some binary operands after benchmarking.

Let me know if you have any questions, happy to discuss those things with you!

—You are receiving this because you were mentioned.Reply to this email directly, view it on GitHub, or unsubscribe.Triage notifications on the go with GitHub Mobile for iOS or Android.

[
{
***@***.***": "http://schema.org",
***@***.***": "EmailMessage",
"potentialAction": {
***@***.***": "ViewAction",
"target": "https://github.com/RRZE-HPC/OSACA/pull/80#issuecomment-981870023",
"url": "https://github.com/RRZE-HPC/OSACA/pull/80#issuecomment-981870023",
"name": "View Pull Request"
},
"description": "View this Pull Request on GitHub",
"publisher": {
***@***.***": "Organization",
"name": "GitHub",
"url": "https://github.com"
}
}
]
